_posts/2017-09-05-Browser Security Features.md

```diff
@@ -0,0 +1,290 @@
+![HPPK error](/img/2017-security-features/hpkp-error.png){: .image height="245px" }
+
+An HPKP header looks like this:
+
+{% highlight http %}
+public-key-pins:pin-sha256="YLh1...uihg=";pin-sha256="9dNi...Dwg=";pin-sha256="Vjs...eWys=";max-age=2592000;includeSubdomains;report-uri="report-uri"
+{% endhighlight %}{: style="scroll: auto" }
+
+The `max-age` directive tells the browser for how long these pins are valid. 
+You can use `report-uri` to get a report when an invalid certificate is used.
+`includesubdomains` makes sure that the policy also applies to your subdomains.
+Finally, there are the actual pins.
+You need to pin at least 2 fingerprints: 1 that should be active at the moment and 1 that isn't.
+
+### What to pin?
+First of all you need to pin at least one of the keys in your certificate chain.
+While you can pin the key of the actual certificate, that might not be the best idea.
+Doing this means that you need to update the keys every time your certificate is changed or risk your users being unable to visit your site.
+Alternatively, you could pin the key for the root certificate of your CA (Certificate Authority).
+While this is a lot safer, it does mean that if your CA, or any of its intermediates is compromised, they could issue valid certificates for your site.
+Finally you have the option to pin the key to the intermediate certificate. 
+Doing so limits the attack surface to that intermediate, while it also allows you to roll out new certificates whenever you need to.
+Of course you can't control when your CA will change their intermediate cert, so that's a danger in its own.
+
+Besides that you also need to have a second key pinned that's *not* in your current certificate chain, otherwise your HPKP header will be ignored.
+Fortunately you don't have to have certificates ready for this.
+It's enough to pin the public key of a CSR (Certificate Signing Request).
+Obviously you can't use the CSR of your current certificate (as that would be valid for this chain), so you'll need to create a backup CSR.
+You'll need to keep this CSR and the associated private key in a secure location, because you don't want these to be compromised together with the original.
+
+### report-uri
+The `report-uri` directive is used to tell the browser where to send reports if it encounters an invalid certificate.
+The browser will POST a JSON message to the URL you specify here.
+If you don't want to implement your own processing of these reports, have a look at [report-uri.io](https://report-uri.io/){: target="_blank" }
+It will process the reports from your site and display the results in a nice format, allowing you to take action when you see something that's wrong.
+
+### Report-only
+Besides the normal CSP header, there's also the Report-only variant: `Public-Key-Pins-Report-Only`.
+This header has the exact same specifications as the normal HPKP, **but** it won't block access to your site if there's no valid pin.
```

Comment:
" as the normal HPKP" should be removed because it was mentioned in the sentence just before this one.